pennylane/plugins/default_qubit.py

```diff
@@ -380,6 +379,10 @@ def apply(self, operation):
         self._state = self.mat_vec_product(A, self._state, wires)
         self._first_operation = False
 
+    def post_apply(self):
+        # generate computational basis samples
+        self.generate_samples()
+
```

Comment:
How come this was only implemented for ``default.qubit``?

Also, would there be any reason not to move ``rotate_basis`` into ``post_apply``?

One of the things HW devices could do is to override the ``generate_samples`` method used in ``post_apply``. One of the advantages of that would be, that actually the main difference between a HW device and a statevector simulator would be is how they generate samples.

Reply:
Would it also perhaps be better to move these into ``pre_measure``? Currently (as there was discussion about this) it might still be ambiguous what is happening in each of the ``pre``/``post`` functions

Reply:
These are good points. I'm starting to think `generate_samples` shouldn't be a surprise to the plugin developer, which it might be if it is within `pre_measure`/`post_apply`. Maybe the plugin needs to explicitly call `generate_samples`, or explicitly provide a `bool` that turns this feature on and off.

Since `post_apply`, `pre_measure` etc. don't really have any definition (there convenience methods for the plugin to overwrite if needed), maybe it's a bad idea to put things there by default?

Reply:
I also think that the `default.qubit` plugin should overwrite `generate_samples` so that samples are only generated if it is not in analytic mode.

Reply:
That's a fair point! In the cases when ``analytic`` is ``False`` and the ``return_type`` is not ``Sample``, ``default.qubit`` shouldn't generate samples